crates/bevy_ecs/src/bundle.rs

```diff
@@ -2118,43 +2128,61 @@ fn sorted_remove<T: Eq + Ord + Copy>(source: &mut Vec<T>, remove: &[T]) {
     });
 }
 
-/// BundleEffectFn allows direct access to the BundleEffect trait without having
+/// [`Effect`] wraps a [`BundleEffect`] into a [`Bundle`] allowing
+/// you to leverage the power of a [`BundleEffect`] without having
 /// to implement your own unsafe implementation of the Bundle trait.
+/// [`BundleEffect`]s get access to [`EntityWorldMut`] on insert allowing for
+/// powerful modifications to the world. [`BundleEffect`]s also power other Bevy
+/// helper bundles like [`crate::spawn::SpawnIter`], [`crate::spawn::SpawnWith`], and [`crate::observer::Observer::bundle`].
+/// You can implement BundleEffect directly or use the implementation of BundleEffect
+/// for `FnOnce(&mut EntityWorldMut)`.
 ///
-/// Bundle Effects get access to [`EntityWorldMut`] on insert allowing for
-/// powerful modifications to the world. Bundle Effects also power other Bevy
-/// helper bundles like [`crate::spawn::SpawnIter`] and [`crate::spawn::SpawnWith`]
-pub struct BundleEffectFn<F>(pub F)
+/// Note: [`crate::component::Component`] Hooks are almost as powerful as
```

Comment:
I think this advice needs workshopping. You *can* do structural changes inside of hooks. The real tradeoff to me IMO is "always needed" vs "optionally needed" behavior. Perf might also be impacted, but I would really really want to see benchmarks before comparing.